compiler/noirc_frontend/src/elaborator/mod.rs

```diff
@@ -1227,11 +1225,12 @@ impl<'context> Elaborator<'context> {
             self.push_err(ResolverError::MutableGlobal { span });
         }
 
-        let (let_statement, _typ) = self.elaborate_let(let_stmt);
+        self.elaborate_global_let(let_stmt, global_id);
 
-        let statement_id = self.interner.get_global(global_id).let_statement;
-        self.interner.get_global_definition_mut(global_id).kind = definition_kind.clone();
-        self.interner.replace_statement(statement_id, let_statement);
+        // Avoid defaulting the types of globals here since they may be used in any function.
+        // Otherwise we may prematurely default to a Field inside the next function if this
+        // global was unused there, even if it is consistently used as a u8 everywhere else.
+        self.type_variables.clear();
```

Comment:
Could we potentially be clearing type variables for non-globals that we do want defaulted?

Reply:
Would it be better to do something more similar to recover_generics? recover_type_variables?

Reply:
The `type_variables` field was originally just in the TypeChecker for keeping track of type variables created in functions, it wasn't really meant to be used outside that context. We could add extra tracking so that functions only default type variables they create by remembering the old size of this field then have a check at the very end of the program for defaulting any globals afterward. That's unneeded though since monomorphization will later catch any globals and either default them or error.

Reply:
Ok thanks for the clarification